Radial Basis Functions Utility

SADPR opened the pull request:
This method calculates the RBF shape function values in one point using as
support the given cloud of points.

Works correctly, not yet optimized (might be slow).

SADPR committed:
Radial Basis Functions Utility:
This method calculates the RBF shape function values in one point using as
support the given cloud of points.

Works correctly, not yet optimized (may be slow).

EduardGomezEscandell commented on kratos/tests/test_radial_basis_functions_utility.py:
```diff
@@ -0,0 +1,105 @@
+    gid_output.ExecuteInitializeSolutionStep()
+    gid_output.PrintOutput()
+    gid_output.ExecuteFinalizeSolutionStep()
+    gid_output.ExecuteFinalize()
```
This function should be a method of `TestRadialBasisFunctionsUtility`.

EduardGomezEscandell commented on kratos/tests/test_radial_basis_functions_utility.py:
```diff
@@ -0,0 +1,105 @@
+            for node in self.main_model_part.Nodes:
+                node.SetSolutionStepValue(KratosMultiphysics.TEMPERATURE, 0, N_container[i])
+                i += 1
+            PostProcess(self.main_model_part, "test_radial_basis_functions_utility_1x1_square")
```
Is `PostProcess` necessary for the test or just for debugging?

If the latter, I would comment it out or remove it altogether.

SADPR replied:
It is not necessary, we can comment it.

EduardGomezEscandell commented on kratos/tests/test_radial_basis_functions_utility.py:
```diff
@@ -0,0 +1,105 @@
+        midpoint[2] = 0.0
+        N_container = KratosMultiphysics.Vector()
+        h = 1 #Shape parameter for the kernel function (radial basis function, inverse multiquadratic)
+        InputShapeParameter = False
```
```suggestion
        input_shape_parameter = False
```
Local variables should be lower case

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+{
+
+    double RadialBasisFunctionsUtility::EvaluateRBF(
+        double x,
```
```suggestion
        const double x,
```

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+        // Build the RBF interpolation matrix and RBF interpolated vector
+        for (std::size_t i_pt = 0; i_pt < n_points; ++i_pt) {
+            for (std::size_t j_pt = 0; j_pt < n_points; ++j_pt) {
+                norm_xij = norm_2(row(rPoints,i_pt)-row(rPoints,j_pt));
```
Stack-allocated variables don't need to be preallocated. Declare `norm_xij` and `norm_X` only when you need them. Then you can also make them `const`.

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+        MathUtils<double>::InvertMatrix(A, A_inv, A_det);
+
+        // Obtain the RBF shape functions 
+        noalias(rN) = prod(A_inv,Phi);
```
Inverting A can be quite expensive, in this case you can use `MathUtils::Solve` and solve the system `A*N = Φ`.

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+        double d = 0; // Total distance of nearest x_i neighbors 
+        double h; // Shape parameter
+        for (std::size_t i_pt = 0; i_pt < n_points; ++i_pt) {
+            double d_nearest_neighbor = 1e30; // Initialize distance to nearest x_i neighbor 
```
```suggestion
            double d_nearest_neighbor = std::numeric_limits<double>::max; // Initialize distance to nearest x_i neighbor 
```

You will have to `#include<limits>`as well.

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+
+        // Find shape parameter http://www.math.iit.edu/~fass/Dolomites.pdf [Hardy]
+        // TODO: Build a matrix with distance norm values to recycle for RBFs evaluation.
+        // Make a function for this 
```
Agree :smiley:

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+        // TODO: Build a matrix with distance norm values to recycle for RBFs evaluation.
+        // Make a function for this 
+        double d = 0; // Total distance of nearest x_i neighbors 
+        double h; // Shape parameter
```
Wait until needed to declare the variable, then you can make it const.

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+        MathUtils<double>::InvertMatrix(A, A_inv, A_det);
+
+        // Obtain the RBF shape functions 
+        noalias(rN) = prod(A_inv,Y);
```
Same advice about solving rather than inverting

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+
+        // Interpolate solution
+        double Phi;
+        double norm_xi;
```
No need for preallocation. Declare them when needed.

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.h:
```diff
@@ -0,0 +1,80 @@
+     * @return double The RBF value
+     */
+    static double EvaluateRBF(
+        double x,
```
```suggestion
        const double x,
```

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+        noalias(rN) = prod(Phi,A_inv);
+
+        // Partition of unity
+        noalias(rN) =  rN/sum(rN);
```
Could this block of code be replaced by the following line?

```C++
    CalculateShapeFunctions(rPoints, rX, h, rN);
```

This way we avoid duplicated code.

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.cpp:
```diff
@@ -0,0 +1,208 @@
+        const array_1d<double,3>& rX,
+        const double h,
+        Vector& rN,
+        Vector Y)
```
```suggestion
        Vector& Y)
```
I think.

EduardGomezEscandell commented on kratos/utilities/radial_basis_functions_utility.h:
```diff
@@ -0,0 +1,80 @@
+        const array_1d<double,3>& rX,
+        const double h,
+        Vector& rN,
+        Vector Y);
```
```suggestion
        Vector& Y);
```
Same as previous comment.

EduardGomezEscandell commented on kratos/tests/test_radial_basis_functions_utility.py:
```diff
@@ -0,0 +1,105 @@
+import os
+
```
```suggestion
```
Unused import

SADPR committed:
Update kratos/tests/test_radial_basis_functions_utility.py

Co-authored-by: Eduard Gómez Escandell <egomez@cimne.upc.edu>

SADPR committed:
Update kratos/utilities/radial_basis_functions_utility.cpp

Co-authored-by: Eduard Gómez Escandell <egomez@cimne.upc.edu>

SADPR committed:
Update kratos/utilities/radial_basis_functions_utility.h

Co-authored-by: Eduard Gómez Escandell <egomez@cimne.upc.edu>

SADPR committed:
Update kratos/utilities/radial_basis_functions_utility.h

Co-authored-by: Eduard Gómez Escandell <egomez@cimne.upc.edu>

SADPR committed:
Update kratos/tests/test_radial_basis_functions_utility.py

Co-authored-by: Eduard Gómez Escandell <egomez@cimne.upc.edu>

SADPR committed:
Changed some declaration of variables (i.e. const double norm_xij, h, etc).
Avoid inverted matrix. (Will be changed to QR).
TODO: Repeated code -> Functions

rubenzorrilla committed:
Merge branch 'core/core-qr-using-dense-qr-decomposition-api' into RadialBasisFunctions

Commit:
Adding KRATOS_API to class declaration

Commit:
Merge branch 'master' into RadialBasisFunctions

rubenzorrilla committed:
Merge branch 'RadialBasisFunctions' of https://github.com/KratosMultiphysics/Kratos into RadialBasisFunctions

rubenzorrilla committed:
QR decomposition

rubenzorrilla committed:
Renaming

rubenzorrilla committed:
Missing rename

rubenzorrilla committed:
Merge branch 'master' into RadialBasisFunctions

rubenzorrilla commented on kratos/utilities/rbf_shape_functions_utility.h:
```diff
@@ -0,0 +1,136 @@
+     * @param h Gaussian radius
+     * @return double The RBF value
+     */
+    static double EvaluateRBF(
+        const double x,
+        const double h);
```
I just noted that we can make this `private`. I'm assuming that you don't need/want to keep it `public` though.

rubenzorrilla committed:
Merge branch 'master' into RadialBasisFunctions

rubenzorrilla committed:
Merge branch 'master' into RadialBasisFunctions

rubenzorrilla committed:
Merge branch 'master' into RadialBasisFunctions

rubenzorrilla committed:
Fix ill-conditioned test

EduardGomezEscandell commented on kratos/utilities/rbf_shape_functions_utility.cpp:
```diff
@@ -0,0 +1,169 @@
+        // Build the RBF interpolation matrix and RBF interpolated vector
+        for (std::size_t i_pt = 0; i_pt < n_points; ++i_pt) {
+            for (std::size_t j_pt = 0; j_pt < n_points; ++j_pt) {
+                const double norm_xij = norm_2(row(rPoints,i_pt)-row(rPoints,j_pt));
```
This looks like it is symmetrical, could it be optimized?

EduardGomezEscandell commented on kratos/utilities/rbf_shape_functions_utility.cpp:
```diff
@@ -0,0 +1,169 @@
+        // Build the RBF interpolation matrix and RBF interpolated vector
+        for (std::size_t i_pt = 0; i_pt < n_points; ++i_pt) {
+            for (std::size_t j_pt = 0; j_pt < n_points; ++j_pt) {
+                norm_xij = norm_2(row(rPoints,i_pt)-row(rPoints,j_pt));
```
Same comment regarding symmetry

EduardGomezEscandell commented on kratos/utilities/rbf_shape_functions_utility.cpp:
```diff
@@ -0,0 +1,169 @@
+    void RBFShapeFunctionsUtility::CalculateShapeFunctions(
+        const Matrix& rPoints,
+        const array_1d<double,3>& rX,
+        const double h,
+        Vector& rN,
+        DenseQRPointerType pDenseQR)
+    {
+        KRATOS_TRY;
+
+        KRATOS_ERROR_IF(h < 1.0e-12) << "Reference distance close to zero." << std::endl;
```
I'd either rename `h` to `reference_distance` or something similar; or change the error message to make it obvious `h` is the issue.

RiccardoRossi replied:
we should correct this, i think that Eduard's point is very valid!

EduardGomezEscandell reviewed (approved):
Found some issues, small enough not to block
